nautobot/utilities/viewsets.py

```diff
@@ -0,0 +1,77 @@
+from nautobot.utilities.permissions import get_permission_for_model
+from nautobot.utilities.views import (
+    NautobotViewSetMixin,
+    ObjectPermissionRequiredMixin,
+    ObjectDetailViewMixin,
+    ObjectDeleteViewMixin,
+    ObjectListViewMixin,
+    ObjectEditViewMixin,
+    BulkImportViewMixin,
+    BulkDeleteViewMixin,
+    BulkEditViewMixin,
+    View,
+)
+from rest_framework.routers import SimpleRouter
+from rest_framework.viewsets import ViewSetMixin
+
+
+class NautobotViewSet(
+    ObjectPermissionRequiredMixin,
+    ObjectDetailViewMixin,
+    ObjectListViewMixin,
+    ObjectEditViewMixin,
+    ObjectDeleteViewMixin,
+    BulkDeleteViewMixin,
+    BulkEditViewMixin,
+    BulkImportViewMixin,
+    ViewSetMixin,
+    View,
+):
+    def __init__(self, *args, **kwargs):
+        if kwargs.get("suffix") == "List":
+            self.action = "view"
+            self.template_name = self.get_template_name("list")
+            self.queryset = self.table_queryset
+        elif kwargs.get("suffix") == "Detail":
+            self.action = "view"
+            self.template_name = self.get_template_name("detail")
+            self.queryset = self.detail_queryset
+        elif kwargs.get("suffix") == "Add":
+            self.action = "add"
+            self.template_name = self.get_template_name("edit")
+            self.queryset = self.detail_queryset
+        elif kwargs.get("suffix") == "Import":
+            self.action = "add"
+            self.template_name = self.get_template_name("bulk_import")
+            self.queryset = self.table_queryset
+        elif kwargs.get("suffix") == "Edit":
+            self.action = "change"
+            self.template_name = self.get_template_name("edit")
+            self.queryset = self.detail_queryset
+        elif kwargs.get("suffix") == "Bulk Edit":
+            self.action = "change"
+            self.template_name = self.get_template_name("bulk_edit")
+            self.queryset = self.table_queryset
+        elif kwargs.get("suffix") == "Delete":
+            self.action = "delete"
+            self.template_name = self.get_template_name("delete")
+            self.queryset = self.detail_queryset
+        elif kwargs.get("suffix") == "Bulk Delete":
+            self.action = "delete"
+            self.template_name = self.get_template_name("bulk_delete")
+            self.queryset = self.table_queryset
+        else:
+            self.action = "view"
+            self.template_name = self.get_template_name("list")
+            self.queryset = self.detail_queryset
+
+        super().__init__(*args, **kwargs)
+
+    def get_required_permission(self):
+        return get_permission_for_model(self.queryset.model, self.action)
+
+
+class NautobotRouter(SimpleRouter, NautobotViewSetMixin):
```

Comment:
`NautobotViewSetMixin` is responsible for concatenating the `routes` variable that the `router `uses.  And all the other mixins have inherited from `NautobotViewSetMixin`. So when I declared `NautobotViewSet`, I thought it wouldnt be necessary to mix in `NautobotViewSetMixin` since it is already done with all the other mixins.

Reply:
Is this because `NautobotViewSetMixin` contains the method `define_routes`? If so when you actually use `NautobotRouter` at least one mixin added to it there should bring along `NautobotViewSetMixin`:

```py
class ProviderNetworkRouter(
     NautobotRouter,
     ObjectDetailViewMixin,
     ObjectListViewMixin,
     ObjectEditViewMixin,
     ObjectDeleteViewMixin,
     BulkDeleteViewMixin,
     BulkImportViewMixin,
     BulkEditViewMixin,
 ):
#...
```

Therefor the only way `self.routes = super().define_routes()` should fail is if you fail to mix in any views to the router which I would presume would be an invalid use case in its current setup.